pkg/storage/unified/sql/db/migrations/migrator.go

```diff
@@ -0,0 +1,24 @@
+package migrations
```

Comment:
Given that we are already making a refactor of the database, can we get rid of the legacy `migrator` package? It has too many issues, it is very poorly tested and has many bad practices in place. I suggest the following:

1. Use the `sqltemplate` package to write migrations.
2. Write two files per migration, one ending in `-up.sql` and one ending in `-down.sql`. A single migration can perform any number of changes, they will be performed in the same transaction.
3. Prepend each file with a `NNN-` number starting at `001-`.
4. Use the `000-` exclusively to create a migrations table, which should have a single column and a single row holding the last committed migration.
5. Do not use `PRIMARY KEY`. Instead, use `CREATE UNIQUE INDEX` and require each column to be `NOT NULL`. This has the same desired effect as a primary key, but increases interoperability.

<details><summary>SQL boilerplate example</summary>
<p>

For the migrations table (note that we use singular in all table names, so we call it `migration`):

```sql
-- file migrations/data/000-migration-down.sql
DROP TABLE {{ .Ident "migration" }};

-- file migrations/data/000-migration-up.sql
CREATE TABLE {{ .Ident "migration: }} ( {{ .Ident "last_migration_committed" }} BIGINT);
```

Example snippet for the first migration.

```sql
-- file migrations/data/001-initial-down.sql
DROP INDEX {{ .Ident "idx_resource_pk" }} ON  {{ .Ident "resource" }};
DROP INDEX {{ .Ident "idx_resource__namespace" }} ON  {{ .Ident "resource" }};
DROP INDEX {{ .Ident "idx_resource__key" }} ON  {{ .Ident "resource" }};
DROP TABLE {{ .Ident "resource" }};
-- all the other code to drop the rest of the things made in -up.sql

-- file migrations/data/001-initial-up.sql
CREATE TABLE {{ .Ident "resource" }} (
    {{ .Ident "guid"}} CHAR(36) NOT NULL,
    {{ .Ident "resource_version"}} BIGINT,

    {{ .Ident "group"}} VARCHAR(190) NOT NULL,
    {{ .Ident "resource_version"}} VARCHAR(190) NOT NULL,
    {{ .Ident "namespace"}} VARCHAR(63) NOT NULL,
    {{ .Ident "name"}} VARCHAR(190) NOT NULL,
    {{ .Ident "value"}} TEXT,
    {{ .Ident "action"}} SMALLINT NOT NULL,
    {{ .Ident "label_set"}} VARCHAR(64)
);

-- use two underscores to separate the table name from the rest of the index name for clarity
CREATE UNIQUE INDEX {{ .Ident "idx_resource__pk" }} ON  {{ .Ident "resource" }} (
    {{ .Ident "guid" }}
);
CREATE INDEX {{ .Ident "idx_resource__namespace" }} ON  {{ .Ident "resource" }} (
    {{ .Ident "namespace" }}
);
CREATE INDEX {{ .Ident "idx_resource__key" }} ON  {{ .Ident "resource" }} (
    {{ .Ident "group" }},
    {{ .Ident "resource" }},
    {{ .Ident "name" }}
);
```

</p>
</details> 

For the Go code, we need to embed the templates and execute them with the same `exec` function in `backend.go`, each of them in a separate transaction. In the same transaction, we need to `SELECT ... FOR UPDATE` from the `migration` table, and then `UPDATE` it after running the migration itself. If calling `SELECT ... FOR UPDATE` returns an error, then run all the migrations, otherwise start from the next migration number.

Reply:
As per our conversation, we will defer this work for a later PR.